kubeadm join segfault

Trying to add a worker node to a working cluster. Increading verbosity doesn't give more hints.

K8s: 1.33
OS: Debian Trixie
Linux: 6.12.48+deb13-amd64
Containerd: containerd github.com/containerd/containerd 1.7.24~ds1 1.7.24~ds1-6+b4

```yaml
apiVersion: kubeadm.k8s.io/v1beta4
kind: JoinConfiguration

discovery:
  bootstrapToken:
    token: "zvcv...5amli"
    apiServerEndpoint: "api.mydomain.tld:6443"
    unsafeSkipCAVerification: true
    caCertHashes: ['sha256:be5089...063c10']
  tlsBootstrapToken: "zvc...5amli"

nodeRegistration:
  name: "node11"
```

```
# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"33", EmulationMajor:"", EmulationMinor:"", MinCompatibilityMajor:"", MinCompatibilityMinor:"", GitVersion:"v1.33.5", GitCommit:"03e764d0394bdff662e960c70d25b3c30d731666", GitTreeState:"clean", BuildDate:"2025-09-09T19:50:45Z", GoVersion:"go1.24.6", Compiler:"gc", Platform:"linux/amd64"}
```

```
# kubeadm join --config ./kubeadm-join.yaml -v=5
I1019 01:22:14.894775    6381 join.go:421] [preflight] found NodeName empty; using OS hostname as NodeName
I1019 01:22:14.894828    6381 joinconfiguration.go:83] loading configuration from "./kubeadm-join.yaml"
I1019 01:22:14.895246    6381 initconfiguration.go:123] detected and using CRI socket: unix:///var/run/containerd/containerd.sock
[preflight] Running pre-flight checks
I1019 01:22:14.895292    6381 preflight.go:93] [preflight] Running general checks
I1019 01:22:14.897683    6381 checks.go:278] validating the existence of file /etc/kubernetes/kubelet.conf
I1019 01:22:14.897699    6381 checks.go:278] validating the existence of file /etc/kubernetes/bootstrap-kubelet.conf
I1019 01:22:14.897707    6381 checks.go:102] validating the container runtime
I1019 01:22:14.897991    6381 checks.go:637] validating whether swap is enabled or not
I1019 01:22:14.898027    6381 checks.go:368] validating the presence of executable losetup
I1019 01:22:14.898046    6381 checks.go:368] validating the presence of executable mount
I1019 01:22:14.898060    6381 checks.go:368] validating the presence of executable cp
I1019 01:22:14.898074    6381 checks.go:514] running all checks
I1019 01:22:14.907378    6381 checks.go:399] checking whether the given node name is valid and reachable using net.LookupHost
I1019 01:22:14.907737    6381 checks.go:603] validating kubelet version
I1019 01:22:14.936744    6381 checks.go:128] validating if the "kubelet" service is enabled and active
I1019 01:22:14.949234    6381 checks.go:201] validating availability of port 10250
I1019 01:22:14.949370    6381 checks.go:278] validating the existence of file /etc/kubernetes/pki/ca.crt
I1019 01:22:14.949380    6381 checks.go:428] validating if the connectivity type is via proxy or direct
I1019 01:22:14.949396    6381 join.go:551] [preflight] Discovering cluster-info
I1019 01:22:14.949408    6381 token.go:72] [discovery] Created cluster-info discovery client, requesting info from "api.mydomain.tld:6443"
I1019 01:22:14.949499    6381 envvar.go:172] "Feature gate default state" feature="ClientsAllowCBOR" enabled=false
I1019 01:22:14.949509    6381 envvar.go:172] "Feature gate default state" feature="ClientsPreferCBOR" enabled=false
I1019 01:22:14.949513    6381 envvar.go:172] "Feature gate default state" feature="InformerResourceVersion" enabled=false
I1019 01:22:14.949517    6381 envvar.go:172] "Feature gate default state" feature="InOrderInformers" enabled=true
I1019 01:22:14.949520    6381 envvar.go:172] "Feature gate default state" feature="WatchListClient" enabled=false
I1019 01:22:14.949709    6381 token.go:230] [discovery] Waiting for the cluster-info ConfigMap to receive a JWS signature for token ID "zvcvj1"
I1019 01:22:14.955640    6381 token.go:127] [discovery] Requesting info from "api.mydomain.tld:6443" again to validate TLS against the pinned public key
I1019 01:22:14.955785    6381 token.go:230] [discovery] Waiting for the cluster-info ConfigMap to receive a JWS signature for token ID "zvcvj1"
I1019 01:22:14.961644    6381 token.go:150] [discovery] Cluster info signature and contents are valid and TLS certificate validates against pinned roots, will use API Server "api.mydomain.tld:6443"
I1019 01:22:14.961655    6381 discovery.go:53] [discovery] Using provided TLSBootstrapToken as authentication credentials for the join process
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x24490a9]

goroutine 1 [running]:
k8s.io/kubernetes/cmd/kubeadm/app/discovery.For({0x0?, 0x0?}, 0xc000562d80)
        k8s.io/kubernetes/cmd/kubeadm/app/discovery/discovery.go:57 +0x169
k8s.io/kubernetes/cmd/kubeadm/app/cmd.(*joinData).TLSBootstrapCfg(0xc0004550a0)
        k8s.io/kubernetes/cmd/kubeadm/app/cmd/join.go:552 +0xd4
k8s.io/kubernetes/cmd/kubeadm/app/cmd.(*joinData).InitCfg(0xc0004550a0)
        k8s.io/kubernetes/cmd/kubeadm/app/cmd/join.go:562 +0x29
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/join.runPreflight({0x2a1f840?, 0xc0004550a0})
        k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/join/preflight.go:102 +0x1e5
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1(...)
        k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:261
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll(...)
        k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:450
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run(0xc0004e4900, {0xc000529530, 0x0, 0x3})
        k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:234 +0x502
k8s.io/kubernetes/cmd/kubeadm/app/cmd.newCmdJoin.func1(0xc00063b900?, {0xc000529530, 0x0, 0x3})
        k8s.io/kubernetes/cmd/kubeadm/app/cmd/join.go:185 +0x85
github.com/spf13/cobra.(*Command).execute(0xc000642608, {0xc0005294d0, 0x3, 0x3})
        github.com/spf13/cobra@v1.8.1/command.go:985 +0xad4
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001fa908)
        github.com/spf13/cobra@v1.8.1/command.go:1117 +0x44f
github.com/spf13/cobra.(*Command).Execute(0x2fb3d40?)
        github.com/spf13/cobra@v1.8.1/command.go:1041 +0x13
k8s.io/kubernetes/cmd/kubeadm/app.Run()
        k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:47 +0xad
main.main()
        k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25 +0x13
```

--- a/cmd/kubeadm/app/cmd/join.go
+++ b/cmd/kubeadm/app/cmd/join.go
@@ -692,7 +692,10 @@ func fetchInitConfigurationFromJoinConfiguration(cfg *kubeadmapi.JoinConfigurati
 	}
 
 	// Create the final KubeConfig file with the cluster name discovered after fetching the cluster configuration
-	_, clusterinfo := kubeconfigutil.GetClusterFromKubeConfig(tlsBootstrapCfg)
+	_, clusterinfo, err := kubeconfigutil.GetClusterFromKubeConfig(tlsBootstrapCfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "the TLS bootstrap kubeconfig is malformed")
+	}
 	tlsBootstrapCfg.Clusters = map[string]*clientcmdapi.Cluster{
 		initConfiguration.ClusterName: clusterinfo,
 	}

--- a/cmd/kubeadm/app/cmd/util/join.go
+++ b/cmd/kubeadm/app/cmd/util/join.go
@@ -56,9 +56,9 @@ func getJoinCommand(kubeConfigFile, token, key string, controlPlane, skipTokenPr
 	}
 
 	// load the default cluster config
-	_, clusterConfig := kubeconfigutil.GetClusterFromKubeConfig(config)
-	if clusterConfig == nil {
-		return "", errors.New("failed to get default cluster config")
+	_, clusterConfig, err := kubeconfigutil.GetClusterFromKubeConfig(config)
+	if err != nil {
+		return "", errors.Wrapf(err, "malformed kubeconfig file: %s", kubeConfigFile)
 	}
 
 	// load CA certificates from the kubeconfig (either from PEM data or by file path)

--- a/cmd/kubeadm/app/cmd/util/join_test.go
+++ b/cmd/kubeadm/app/cmd/util/join_test.go
@@ -133,7 +133,7 @@ func TestGetJoinCommand(t *testing.T) {
 			kubeConfig:   &clientcmdapi.Config{},
 			token:        "test-token",
 			expectError:  true,
-			errorMessage: "failed to get default cluster config",
+			errorMessage: "the current context is invalid",
 		},
 		{
 			name:         "Error when CA certificate is invalid",

--- a/cmd/kubeadm/app/discovery/discovery.go
+++ b/cmd/kubeadm/app/discovery/discovery.go
@@ -21,6 +21,7 @@ import (
 
 	clientset "k8s.io/client-go/kubernetes"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	bootstrapapi "k8s.io/cluster-bootstrap/token/api"
 	"k8s.io/klog/v2"
 
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
@@ -51,7 +52,10 @@ func For(client clientset.Interface, cfg *kubeadmapi.JoinConfiguration) (*client
 	if len(cfg.Discovery.TLSBootstrapToken) != 0 {
 		klog.V(1).Info("[discovery] Using provided TLSBootstrapToken as authentication credentials for the join process")
 
-		_, clusterinfo := kubeconfigutil.GetClusterFromKubeConfig(config)
+		_, clusterinfo, err := kubeconfigutil.GetClusterFromKubeConfig(config)
+		if err != nil {
+			return nil, errors.Wrapf(err, "malformed kubeconfig in the %s ConfigMap", bootstrapapi.ConfigMapClusterInfo)
+		}
 		return kubeconfigutil.CreateWithToken(
 			clusterinfo.Server,
 			kubeadmapiv1.DefaultClusterName,

--- a/cmd/kubeadm/app/discovery/file/file.go
+++ b/cmd/kubeadm/app/discovery/file/file.go
@@ -52,9 +52,9 @@ func ValidateConfigInfo(config *clientcmdapi.Config, discoveryTimeout time.Durat
 	if len(config.Clusters) < 1 {
 		return nil, errors.New("the provided kubeconfig file must have at least one Cluster defined")
 	}
-	currentClusterName, currentCluster := kubeconfigutil.GetClusterFromKubeConfig(config)
-	if currentCluster == nil {
-		return nil, errors.New("the provided kubeconfig file must have a unnamed Cluster or a CurrentContext that specifies a non-nil Cluster")
+	currentClusterName, currentCluster, err := kubeconfigutil.GetClusterFromKubeConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "the provided kubeconfig file is malformed")
 	}
 	if err := clientcmd.Validate(*config); err != nil {
 		return nil, err
@@ -124,7 +124,10 @@ func ValidateConfigInfo(config *clientcmdapi.Config, discoveryTimeout time.Durat
 		return config, nil
 	}
 
-	_, refreshedCluster := kubeconfigutil.GetClusterFromKubeConfig(refreshedBaseKubeConfig)
+	_, refreshedCluster, err := kubeconfigutil.GetClusterFromKubeConfig(refreshedBaseKubeConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "malformed kubeconfig in the %s ConfigMap", bootstrapapi.ConfigMapClusterInfo)
+	}
 	if currentCluster.Server != refreshedCluster.Server {
 		klog.Warningf("[discovery] the API Server endpoint %q in use is different from the endpoint %q which defined in the %s ConfigMap", currentCluster.Server, refreshedCluster.Server, bootstrapapi.ConfigMapClusterInfo)
 	}

--- a/cmd/kubeadm/app/discovery/token/token.go
+++ b/cmd/kubeadm/app/discovery/token/token.go
@@ -103,9 +103,9 @@ func retrieveValidatedConfigInfo(client clientset.Interface, cfg *kubeadmapi.Dis
 		return nil, errors.Wrapf(err, "couldn't parse the kubeconfig file in the %s ConfigMap", bootstrapapi.ConfigMapClusterInfo)
 	}
 
-	// The ConfigMap should contain a single cluster
-	if len(insecureConfig.Clusters) != 1 {
-		return nil, errors.Errorf("expected the kubeconfig file in the %s ConfigMap to have a single cluster, but it had %d", bootstrapapi.ConfigMapClusterInfo, len(insecureConfig.Clusters))
+	_, _, err = kubeconfigutil.GetClusterFromKubeConfig(insecureConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "malformed kubeconfig in the %s ConfigMap", bootstrapapi.ConfigMapClusterInfo)
 	}
 
 	// If no TLS root CA pinning was specified, we're done

--- a/cmd/kubeadm/app/discovery/token/token_test.go
+++ b/cmd/kubeadm/app/discovery/token/token_test.go
@@ -77,10 +77,12 @@ users: null
 		name                     string
 		tokenID                  string
 		tokenSecret              string
+		currentContextCluster    string
 		cfg                      *kubeadmapi.Discovery
 		configMap                *fakeConfigMap
 		delayedJWSSignaturePatch bool
 		expectedError            bool
+		expectedErrorString      string
 	}{
 		{
 			// This is the default behavior. The JWS signature is patched after the cluster-info ConfigMap is created
@@ -129,6 +131,24 @@ users: null
 				name: bootstrapapi.ConfigMapClusterInfo,
 				data: nil,
 			},
+		},
+		{
+			name:        "invalid: the kubeconfig in the configmap has the wrong current context",
+			tokenID:     "123456",
+			tokenSecret: "abcdef1234567890",
+			cfg: &kubeadmapi.Discovery{
+				BootstrapToken: &kubeadmapi.BootstrapTokenDiscovery{
+					Token:        "123456.abcdef1234567890",
+					CACertHashes: []string{caCertHash},
+				},
+			},
+			configMap: &fakeConfigMap{
+				name: bootstrapapi.ConfigMapClusterInfo,
+				data: nil,
+			},
+			currentContextCluster: "foo",
+			expectedError:         true,
+			expectedErrorString:   `malformed kubeconfig in the cluster-info ConfigMap: no matching cluster for the current context: token-bootstrap-client@somecluster`,
 		},
 		{
 			name:        "invalid: token format is invalid",
@@ -216,6 +236,13 @@ users: null
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
 			kubeconfig := buildSecureBootstrapKubeConfig("127.0.0.1", []byte(caCert), "somecluster")
+			if len(test.currentContextCluster) > 0 {
+				currentContext := kubeconfig.Contexts[kubeconfig.CurrentContext]
+				if currentContext == nil {
+					t.Fatal("unexpected nil current context")
+				}
+				currentContext.Cluster = test.currentContextCluster
+			}
 			kubeconfigBytes, err := clientcmd.Write(*kubeconfig)
 			if err != nil {
 				t.Fatalf("cannot marshal kubeconfig %v", err)
@@ -267,8 +294,13 @@ users: null
 				t.Errorf("expected error %v, got %v, error: %v", test.expectedError, err != nil, err)
 			}
 
-			// Return if an error is expected
 			if err != nil {
+				if len(test.expectedErrorString) > 0 && test.expectedErrorString != err.Error() {
+					t.Fatalf("expected error string: %s, got: %s",
+						test.expectedErrorString, err.Error())
+				}
+
+				// Return if an error is expected
 				return
 			}
 

--- a/cmd/kubeadm/app/util/kubeconfig/kubeconfig.go
+++ b/cmd/kubeadm/app/util/kubeconfig/kubeconfig.go
@@ -104,17 +104,21 @@ func WriteToDisk(filename string, kubeconfig *clientcmdapi.Config) error {
 }
 
 // GetClusterFromKubeConfig returns the default Cluster of the specified KubeConfig
-func GetClusterFromKubeConfig(config *clientcmdapi.Config) (string, *clientcmdapi.Cluster) {
+func GetClusterFromKubeConfig(config *clientcmdapi.Config) (string, *clientcmdapi.Cluster, error) {
 	// If there is an unnamed cluster object, use it
 	if config.Clusters[""] != nil {
-		return "", config.Clusters[""]
+		return "", config.Clusters[""], nil
 	}
 
 	currentContext := config.Contexts[config.CurrentContext]
 	if currentContext != nil {
-		return currentContext.Cluster, config.Clusters[currentContext.Cluster]
+		if config.Clusters[currentContext.Cluster] != nil {
+			return currentContext.Cluster, config.Clusters[currentContext.Cluster], nil
+		}
+		return "", nil, errors.Errorf("no matching cluster for the current context: %s", config.CurrentContext)
 	}
-	return "", nil
+
+	return "", nil, errors.Errorf("the current context is invalid: %s", config.CurrentContext)
 }
 
 // HasAuthenticationCredentials returns true if the current user has valid authentication credentials for

--- a/cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go
+++ b/cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go
@@ -351,17 +351,49 @@ func TestGetClusterFromKubeConfig(t *testing.T) {
 		config              *clientcmdapi.Config
 		expectedClusterName string
 		expectedCluster     *clientcmdapi.Cluster
+		expectedError       bool
 	}{
 		{
-			name: "cluster is empty",
+			name: "an existing cluster with an empty name is returned directly",
 			config: &clientcmdapi.Config{
-				CurrentContext: "kubernetes",
+				Clusters: map[string]*clientcmdapi.Cluster{
+					"": {Server: "http://foo:8080"},
+				},
+			},
+			expectedClusterName: "",
+			expectedCluster: &clientcmdapi.Cluster{
+				Server: "http://foo:8080",
+			},
+		},
+		{
+			name: "the current context is invalid",
+			config: &clientcmdapi.Config{
+				CurrentContext: "foo",
+				Contexts: map[string]*clientcmdapi.Context{
+					"bar": {AuthInfo: "bar", Cluster: "bar"},
+				},
 			},
 			expectedClusterName: "",
 			expectedCluster:     nil,
+			expectedError:       true,
 		},
 		{
-			name: "cluster and currentContext are not empty",
+			name: "no matching cluster for the current context",
+			config: &clientcmdapi.Config{
+				CurrentContext: "foo",
+				Contexts: map[string]*clientcmdapi.Context{
+					"foo": {AuthInfo: "bar", Cluster: "bar"},
+				},
+				Clusters: map[string]*clientcmdapi.Cluster{
+					"baz": {Server: "https://bar:16443"},
+				},
+			},
+			expectedClusterName: "",
+			expectedCluster:     nil,
+			expectedError:       true,
+		},
+		{
+			name: "valid current context and cluster",
 			config: &clientcmdapi.Config{
 				CurrentContext: "foo",
 				Contexts: map[string]*clientcmdapi.Context{
@@ -378,30 +410,19 @@ func TestGetClusterFromKubeConfig(t *testing.T) {
 				Server: "http://foo:8080",
 			},
 		},
-		{
-			name: "cluster is not empty and currentContext is not in contexts",
-			config: &clientcmdapi.Config{
-				CurrentContext: "foo",
-				Contexts: map[string]*clientcmdapi.Context{
-					"bar": {AuthInfo: "bar", Cluster: "bar"},
-				},
-				Clusters: map[string]*clientcmdapi.Cluster{
-					"foo": {Server: "http://foo:8080"},
-					"bar": {Server: "https://bar:16443"},
-				},
-			},
-			expectedClusterName: "",
-			expectedCluster:     nil,
-		},
 	}
 	for _, rt := range tests {
 		t.Run(rt.name, func(t *testing.T) {
-			clusterName, cluster := GetClusterFromKubeConfig(rt.config)
+			clusterName, cluster, err := GetClusterFromKubeConfig(rt.config)
 			if clusterName != rt.expectedClusterName {
 				t.Errorf("got cluster name = %s, expected %s", clusterName, rt.expectedClusterName)
 			}
 			if !reflect.DeepEqual(cluster, rt.expectedCluster) {
 				t.Errorf("got cluster = %+v, expected %+v", cluster, rt.expectedCluster)
+			}
+			if (err != nil) != rt.expectedError {
+				t.Errorf("expected error: %v, got: %v, error: %v",
+					rt.expectedError, err != nil, err)
 			}
 		})
 	}

